src/bio.c

```diff
@@ -278,6 +293,158 @@ void *bioProcessBackgroundJobs(void *arg) {
             if (job_type == BIO_CLOSE_AOF) close(job->fd_args.fd);
         } else if (job_type == BIO_LAZY_FREE) {
             job->free_args.free_fn(job->free_args.free_args);
+        } else if (job_type == BIO_SAVE_TO_DISK) {
+            int usemark;
+            char lastbytes[RDB_EOF_MARK_SIZE], buf[PROTO_IOBUF_LEN], eofmark[RDB_EOF_MARK_SIZE];
+            ssize_t readlen, nread, nwritten;
+            off_t left = 0, repl_transfer_last_fsync_off = 0;
+            int error = 0, main_thread_abort = 0, received_ping = 0;
+            connection *conn = job->save_to_disk_args.conn;
+            int dest_fd = job->save_to_disk_args.dest_fd;
+
+            /* Put the socket in blocking mode to simplify RDB transfer.
+             * We'll restore it when the RDB is received. */
+            connBlock(conn);
+            connRecvTimeout(conn, server.repl_syncio_timeout * 1000);
+            do {
```

Comment:
why do we need two loops in the code?

Reply:
We are basically imitating what `readSyncBulkPayload` does for normal replication. We try to read the bulk payload length from the primary. Ideally one pass would be enough (no need for loop), but the primary is sometimes not fast enough in sending the length, so it periodically sends pings (newlines) until it's ready. We have to keep looping while we receive these pings.

The second loop goes on until the primary finishes sending all the data (we see an EOF or the amount we read is equal to the previously passed payload length).

So the first loop is conditioned on receiving pings, the second one on sync completion.